trio/_core/tests/test_windows.py

```diff
@@ -45,5 +51,135 @@
+            ffi.NULL,  # no security attributes
+            FileFlags.OPEN_EXISTING,
+            FileFlags.FILE_FLAG_OVERLAPPED,
+            ffi.NULL,  # no template file
+        )
+        if handle == INVALID_HANDLE_VALUE:  # pragma: no cover
+            raise_winerror()
+
+        try:
+            with memoryview(buffer) as buffer_view:
+
+                async def read_region(start, end):
+                    await _core.readinto_overlapped(
+                        handle,
+                        buffer_view[start:end],
+                        start,
+                    )
+
+                _core.register_with_iocp(handle)
+                async with _core.open_nursery() as nursery:
+                    for start in range(0, 4096, 512):
+                        nursery.start_soon(read_region, start, start + 512)
+
+                assert buffer == data
+
+                with pytest.raises(BufferError):
+                    await _core.readinto_overlapped(handle, b"immutable")
+        finally:
+            kernel32.CloseHandle(handle)
+
+
+@contextmanager
+def pipe_with_overlapped_read():
+    from asyncio.windows_utils import pipe
+    import msvcrt
+
+    read_handle, write_handle = pipe(overlapped=(True, False))
+    try:
+        write_fd = msvcrt.open_osfhandle(write_handle, 0)
+        yield os.fdopen(write_fd, "wb", closefd=False), read_handle
+    finally:
+        kernel32.CloseHandle(ffi.cast("HANDLE", read_handle))
+        kernel32.CloseHandle(ffi.cast("HANDLE", write_handle))
+
+
+def test_forgot_to_register_with_iocp():
+    with pipe_with_overlapped_read() as (write_fp, read_handle):
+        with write_fp:
+            write_fp.write(b"test\n")
+
+        left_run_yet = False
+        nursery = None
+
+        async def main():
+            nonlocal nursery
+            target = bytearray(1)
+            try:
+                async with _core.open_nursery() as nursery:
+                    nursery.start_soon(
+                        _core.readinto_overlapped,
+                        read_handle,
+                        target,
+                        name="xyz"
+                    )
+                    await wait_all_tasks_blocked()
+                    nursery.cancel_scope.cancel()
+            finally:
+                # Run loop is exited without unwinding running tasks, so
+                # we don't get here until the main() coroutine is GC'ed
+                assert left_run_yet
+
+        with pytest.raises(_core.TrioInternalError) as exc_info:
+            _core.run(main)
+        left_run_yet = True
+        assert "Failed to cancel overlapped I/O in xyz " in str(exc_info.value)
+        assert "forget to call register_with_iocp()?" in str(exc_info.value)
+
+        # Suppress the Nursery.__del__ assertion about dangling children,
+        # for both the nursery we created and the system nursery
+        nursery._children.clear()
+        nursery.parent_task.parent_nursery._children.clear()
```

Comment:
Since this is just hiding some messages printed by `__del__`, I think it'd be simpler to let them be printed, and let pytest hide them?

We should use `gc_collect_harder()` though to make sure that they do get printed during this test, so pytest will attach them to the right place.